mdtoc.go

```diff
@@ -1,7 +1,71 @@
 package mdtoc
 
-import "io"
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+)
 
-func Generate(input io.Reader, output io.Writer) {
-	io.Copy(output, input)
+const headerFormat = "- [%s](#%s)"
+
+func writeHeader(output io.Writer, level int, header string) {
+	// TODO: Handle special characters on header
+	line := fmt.Sprintf(headerFormat, header, strings.ToLower(header))
+	// TODO: Handle output writing errors
+	// TODO: Handle when header has trailing spaces
+	output.Write([]byte(line + "\n"))
+}
+
+func parseHeader(parsed []string) (int, string) {
+	var level int
+	for parsed[level] == "#" {
+		level += 1
+	}
+	return level, strings.Join(parsed[level:], "")
+}
+
+func Generate(input io.Reader, output io.Writer) error {
+	headerStart := []byte("<!-- mdtocstart -->")
+	tocHeader := []byte("# Table of Contents")
+	headerEnd := []byte("<!-- mdtocend -->")
+	scanner := bufio.NewScanner(input)
+	var original bytes.Buffer
+	var wroteHeader bool
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		_, err := original.Write([]byte(line + "\n"))
+		if err != nil {
+			return err
+		}
+		parsed := strings.Split(line, "#")
```

Comment:
It was a simple example. To get runes, just iterate over the string with `for Idx, r := range line {}` then golang handles that automagically for you. You can make it explicit importing `utf8` and using `DecodeRuneInString`.

It's obscure because you're spliting the entire line only for counting.. and then to count you use a `for`  loop anyway in a different function... This could be solved one way with a simple for in a proper 'parseHeader' function.

Think: you slice the line by adjacent runes. Why? You can just count...